[stdlib] Allow a default for optional interpolations

natecook1000 opened the pull request:
This adds an `appendInterpolation` overload to `DefaultStringInterpolation` that includes a parameter for providing a default string when the value to interpolate is `nil`. This allows this kind of usage:

```swift
let age: Int? = nil
print("Your age is \(age, default: "timeless")")
// Prints "Your age is timeless"
```

Includes an additional fixit when optional values are interpolated to use this `default:` parameter.

natecook1000 committed:
Allow a default for optional interpolations

This adds an `appendInterpolation` overload to `DefaultStringInterpolation`
that includes a parameter for providing a default string when the value
to interpolate is `nil`. This allows this kind of usage:

    let age: Int? = nil
    print("Your age is \(age, default: "timeless")")
    // Prints "Your age is timeless"

Includes an additional fixit when optional values are interpolated to
use this `default:` parameter.

natecook1000 commented on lib/Sema/MiscDiagnostics.cpp:
```diff
@@ -5688,6 +5688,11 @@ static void diagnoseUnintendedOptionalBehavior(const Expr *E,
 
       if (kind == UnintendedInterpolationKind::Optional) {
+        // Suggest using a default interpolation value parameter.
+        Ctx.Diags.diagnose(arg->getLoc(), diag::default_optional_parameter)
```
@beccadax Do you know if it would be possible to only provide this fixit when the string interpolation type is `DefaultStringInterpolation`?

beccadax replied:
If you had the caller extract the `ConcreteDeclRef` and pass it in, then yes, that ought to be possible.

As an aside, I'm thinking we might not want to suggest *both* `\(_:default:)` *and* `??`, as they mean nearly the same thing. Maybe `??` if it's an `Optional<String>` and `\(_:default:)` otherwise?

natecook1000 replied:
I updated the logic so that the new fixit is only provided for `DefaultStringInterpolation` types, and isn't provided for optional strings, since it's the same functionality as the preferred `??`. I left in the `??` suggestion for other optional types, since there could be some types or situations where a default value of that same type still makes sense.

(I also moved the `String(describing:)` fix-it to be the last suggestion, since it feels like "fix this warning" should be preferred over "silence this warning".)

beccadax commented on stdlib/public/core/StringInterpolation.swift:
```diff
@@ -197,6 +197,46 @@ public struct DefaultStringInterpolation: StringInterpolationProtocol, Sendable
+  }
+
+  @backDeployed(before: SwiftStdlib 6.2)
+  @_disfavoredOverload
```
Instead of using `@_disfavoredOverload` for these, you should be able to provide a fourth `T: TextOutputStreamable, T: CustomStringConvertible` overload. That's how regular `appendInterpolation(_:)` works. (The type checker *really* penalizes disfavored overloads, so it's usually better to avoid it if you can.)

Also, consider making these `@inlinable`. When I originally implemented `DefaultStringInterpolation`, I did some interpolation performance benchmarking (the test should still be in the benchmark suite) and found that inlining really matters for it.

natecook1000 replied:
Oh, I didn't realize! I thought I was helping out by only providing three overloads instead of four. 🙃 I'll look at adding `@inlinable`, too. Thanks!

natecook1000 committed:
Merge branch 'main' into string-interp-default

natecook1000 committed:
Move `String(describing:)` fix to be last

This fix-it suppresses the string interpolation warning, but doesn't
provide any additional affordance for dealing with the problem, so it
shouldn't be the first suggestion.

natecook1000 committed:
Limit when `\(_:default:)` is recommended

Modify fix-its so the new interpolation API is only recommended
for non-strings (since `??` is a more fluent resolution for strings)
and when used in a `DefaultStringInterpolation` type (since other
interpolation types won't have this API).

xwu commented:
@swift-ci Please smoke test

natecook1000 committed:
Add proper overload for string interpolation

natecook1000 commented:
@swift-ci Please smoke test

natecook1000 committed:
Update IDE completion tests

natecook1000 committed:
Update other tests with new APIs

natecook1000 commented:
@swift-ci Please smoke test

natecook1000 committed:
One more slight test update

natecook1000 commented:
@swift-ci Please smoke test

xwu commented:
@swift-ci Please build toolchain

xwu commented:
Please build macOS toolchain

xwu commented:
@swift-ci Please build macOS toolchain

natecook1000 commented:
@swift-ci Please build toolchain macOS platform

natecook1000 committed:
Let default parameter take `some StringProtocol`

natecook1000 committed:
Add documentation

natecook1000 committed:
Relax related wording.

natecook1000 committed:
Merge branch 'main' into string-interp-default

natecook1000 committed:
Update api-digester test for new format

natecook1000 commented:
@swift-ci Please smoke test

natecook1000 commented:
@swift-ci Please smoke test Linux platform

natecook1000 commented:
@swift-ci Please smoke test Linux platform

amartini51 commented on stdlib/public/core/StringInterpolation.swift:
```diff
@@ -197,6 +197,128 @@ public struct DefaultStringInterpolation: StringInterpolationProtocol, Sendable
+  /// Interpolates the given optional value's textual representation, or the
+  /// specified default string, into the string literal being created.
+  ///
+  /// You don't need to call this method directly. It is used by the compiler
```
Here and in the other places this phrasing appears, missing a contraction:

```suggestion
  /// You don't need to call this method directly. It's used by the compiler
```

amartini51 commented on stdlib/public/core/StringInterpolation.swift:
```diff
@@ -197,6 +197,128 @@ public struct DefaultStringInterpolation: StringInterpolationProtocol, Sendable
+  ///
+  /// You don't need to call this method directly. It is used by the compiler
+  /// when interpreting string interpolations where you provide a `default`
+  /// parameter.
```
The code listing would benefit from a sentence introducing it — that might also give you an earlier place to frame the fact that the default interpolation value is used when the value's `nil`.

amartini51 reviewed (approved):
Looks good — left minor feedback and a suggestion.

natecook1000 committed:
Improve string interpolation docs

amartini51 commented:
Changes in 3973a36 look good to me!

Azoy commented on stdlib/public/core/StringInterpolation.swift:
```diff
@@ -197,6 +197,132 @@ public struct DefaultStringInterpolation: StringInterpolationProtocol, Sendable
+  ///   - value: The value to include in a string interpolation, if non-`nil`.
+  ///   - default: The string to include if `value` is `nil`.
+  @backDeployed(before: SwiftStdlib 6.2)
+  @inlinable
```
Any reason for `@backDeployed`+`@inlinable` instead of just `@_alwaysEmitIntoClient`?

natecook1000 replied:
It seemed like the right set, presuming that `@inlinable` has different behavior than `@_alwaysEmit...`. Do we have guidance about which to prefer/when between inlinable and always-emit?

Azoy replied:
These are very simple functions and I'm unsure if we want to commit these as "ABI". With AEIC, the stdlib doesn't take the ABI hit and clients who don't use it don't pay for it either.

natecook1000 committed:
Switch to always-emit

natecook1000 committed:
Merge branch 'main' into string-interp-default

natecook1000 committed:
Update ABI/API test fixtures

natecook1000 commented:
@swift-ci Please smoke test Windows platform
